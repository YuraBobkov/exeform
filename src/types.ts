import { ChangeEvent } from 'react';

export type Touched = {
  [path: string]: boolean;
};

export type Errors = {
  [path: string]: string;
};

export type Validate<Values> = (values: Values) => Errors;

export type Options<Values> = {
  initialValues: Values;
  onSubmit(values: Values): void;
  validate?: Validate<Values>;
};

export type FieldOptions = {
  type: 'generic' | 'checkbox';
};

export type Field = {
  name: string;
  value: any;
  checked?: boolean;
  onChange: ({ target }: ChangeEvent<any>) => void;
  onBlur: () => void;
};

export type Listener = () => void;

export type Unsubscribe = () => void;

export type Form<Values = any> = {
  values: Values;
  touched: Touched;
  errors: Errors;
  isValid: boolean;
  setFieldValue(name: string, value: any): void;
  setFieldTouched(name: string, isTouched?: boolean): void;
  setFieldError(name: string, message: string): void;
  touchAllFields(): void;
  submit(): void;
  subscribe(listener: Listener): Unsubscribe;
};
