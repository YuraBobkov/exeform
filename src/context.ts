import { createContext } from 'react';

import { FormType } from './types';

export const FormContext = createContext<FormType | null>(null);

export const FormProvider = FormContext.Provider;
