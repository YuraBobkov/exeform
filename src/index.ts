export * from './types';

export { default as useForm } from './useForm';

export { default as useFormContext } from './useFormContext';

export { default as useField } from './useField';

export { default as useCheckboxField } from './useCheckboxField';

export { get } from './utils';
