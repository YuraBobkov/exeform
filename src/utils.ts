import { Errors, Touched } from './types';

const isNull = (value: any) => value === null;

const isObject = (value: any): value is { [key: string]: any } =>
  !isNull(value) && typeof value === 'object';

const isNumber = (value: any): value is number => typeof value === 'number';

const isString = (value: any): value is string => typeof value === 'string';

const parseName = (name: string) =>
  name.split('.').reduce((acc: (string | number)[], key) => {
    const match = key.match(/(.+)\[(\d+)\]$/);

    if (match) {
      acc.push(match[1], Number(match[2]));
    } else {
      acc.push(key);
    }

    return acc;
  }, []);

export const get = <Values>(values: Values, name: string) => {
  const getInner = (keys: (string | number)[], valuesInner: any): any => {
    if (!keys.length) {
      return valuesInner;
    }

    const [key, ...restKeys] = keys;
    const value = valuesInner[key];

    if (Array.isArray(value) || isObject(value)) {
      return getInner(restKeys, value);
    }

    return restKeys.length ? undefined : value;
  };

  return getInner(parseName(name), values);
};

export const set = <Values>(
  values: Values,
  name: string,
  value: any,
): Values => {
  const setInner = (keys: (string | number)[], valuesInner: any): any => {
    if (!keys.length) {
      return value;
    }

    const [key, ...restKeys] = keys;

    if (!isObject(valuesInner) && !Array.isArray(valuesInner)) {
      valuesInner = isNumber(key) ? [] : {};
    }

    if (isString(key) && isObject(valuesInner)) {
      return {
        ...valuesInner,
        [key]: setInner(restKeys, valuesInner[key]),
      };
    }

    const arr = valuesInner.slice();

    arr[key] = setInner(restKeys, valuesInner[key]);

    return arr;
  };

  return setInner(parseName(name), values);
};

export const getTouched = (errors: Errors) =>
  Object.keys(errors).reduce((acc: Touched, key) => {
    acc[key] = true;

    return acc;
  }, {});

export const omitOne = <T extends any>(values: T, key: string): T => {
  const copy = { ...values };

  delete copy[key];

  return copy;
};

export const isEmpty = (value: {}) => Boolean(Object.keys(value).length);

export const pick = <T extends any>(values: T, keys: string[]) =>
  keys.reduce((acc, key) => {
    acc[key] = values[key];

    return acc;
  }, {} as T);
