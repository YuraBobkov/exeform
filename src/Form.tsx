import React, { useCallback, PropsWithChildren } from 'react';

import { FormProvider } from './context';
import { Form as FormType, FormProps } from './types';

type Props<Values> = FormProps<Values> & {
  form: FormType<Values>;
};

const Form = <Values,>(props: PropsWithChildren<Props<Values>>) => {
  const { form, onSubmit, ...rest } = props;

  const handleSubmit = useCallback(
    (event) => {
      event.preventDefault();
      form.touchAllFields();

      if (form.isValid) {
        onSubmit(form.values);
      }
    },
    [onSubmit, form],
  );

  return (
    <FormProvider value={form}>
      <form {...rest} noValidate onSubmit={handleSubmit} />
    </FormProvider>
  );
};

export default Form;
