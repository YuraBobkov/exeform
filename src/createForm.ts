import { set, getTouched, omitOne, isEmpty, pick } from './utils';
import { Options, Form, Listener, Touched, Errors } from './types';

const defaultValidate = () => ({});

const createForm = <Values>(options: Options<Values>): Form<Values> => {
  const { validate = defaultValidate, initialValues, onSubmit } = options;
  const listeners: Listener[] = [];
  let values = initialValues;
  let touched: Touched = {};
  let errors = validate(initialValues);
  let externalErrors: Errors = {};

  const notify = () => {
    const currentListeners = listeners.slice();

    currentListeners.forEach((listener) => listener());
  };

  const setFieldTouched = (name: string, isTouched = true) => {
    touched = isTouched
      ? { ...touched, [name]: isTouched }
      : omitOne(touched, name);
  };

  const touchAllFields = () => {
    touched = getTouched(errors);

    notify();
  };

  const getErrors = () => ({ ...errors, ...externalErrors });

  const isValid = () => isEmpty(getErrors());

  return {
    get values() {
      return values;
    },

    get touched() {
      return touched;
    },

    get errors() {
      return getErrors();
    },

    get isValid() {
      return isValid();
    },

    setFieldValue(name, value) {
      values = set(values, name, value);
      errors = validate(values);
      externalErrors = omitOne(externalErrors, name);
      touched = pick(touched, Object.keys(getErrors()));

      notify();
    },

    setFieldTouched(name, isTouched) {
      setFieldTouched(name, isTouched);
      notify();
    },

    setFieldError(name, message) {
      externalErrors = { ...externalErrors, [name]: message };

      setFieldTouched(name);
      notify();
    },

    touchAllFields,

    submit() {
      touchAllFields();

      if (isValid()) {
        onSubmit(values);
      }
    },

    subscribe(fn) {
      listeners.push(fn);

      const unsubscribe = () => {
        const index = listeners.indexOf(fn);

        listeners.splice(index, 1);
      };

      return unsubscribe;
    },
  };
};

export default createForm;
