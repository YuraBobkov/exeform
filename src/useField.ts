import { useCallback, useEffect, useState, ChangeEvent } from 'react';

import { get } from './utils';
import useFormContext from './useFormContext';
import { FieldOptions, Field } from './types';

const useField = (name: string, options?: FieldOptions) => {
  const form = useFormContext();
  const [value, setValue] = useState(get(form.values, name));
  const [error, setError] = useState(form.errors[name] || null);
  const [touched, setTouched] = useState(form.touched[name] || false);
  const type = options?.type || 'generic';
  const isCheckbox = type === 'checkbox';

  useEffect(() => {
    const unsubscribe = form.subscribe(() => {
      setValue(get(form.values, name));
      setError(form.errors[name] || null);
      setTouched(form.touched[name] || false);
    });

    return unsubscribe;
  }, [name, form]);

  const setValueHelper = useCallback(
    (value) => {
      form.setFieldValue(name, value);
    },
    [form, name],
  );

  const setTouchedHelper = useCallback(
    (isTouched = true) => {
      form.setFieldTouched(name, isTouched);
    },
    [form, name],
  );

  const setErrorHelper = useCallback(
    (message: string) => {
      form.setFieldError(name, message);
    },
    [form, name],
  );

  const onChange = useCallback(
    ({ target }: ChangeEvent<any>) => {
      setValueHelper(isCheckbox ? target.checked : target.value);
    },
    [isCheckbox, setValueHelper],
  );

  const onBlur = useCallback(() => {
    setTouchedHelper();
  }, []);

  const getGenericField = (): Field => ({ name, value, onChange, onBlur });

  const getCheckboxField = (): Field => ({
    ...getGenericField(),
    value: undefined,
    checked: value,
  });

  return {
    field: isCheckbox ? getCheckboxField() : getGenericField(),
    meta: { error, touched },
    helpers: {
      setValue: setValueHelper,
      setTouched: setTouchedHelper,
      setError: setErrorHelper,
    },
  };
};

export default useField;
