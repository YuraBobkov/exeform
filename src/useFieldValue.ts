import useField from './useField';

const useFieldValue = (name: string) => useField(name).field.value;

export default useFieldValue;
