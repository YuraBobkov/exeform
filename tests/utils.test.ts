import { get } from '../src/utils';

describe('get', () => {
  const values = {
    id: 42,
    info: {
      name: 'John',
    },
    commentsIds: [56, 78],
    comments: [
      {
        id: 56,
        text: 'foo',
      },
      {
        id: 78,
        text: 'bar',
      },
    ],
  };

  it('should return value', () => {
    expect(get(values, 'id')).toBe(values.id);
  });

  it('should return nested value', () => {
    expect(get(values, 'info.name')).toBe(values.info.name);
  });

  it('should return value from array', () => {
    expect(get(values, 'commentsIds.0')).toBe(values.commentsIds[0]);
    expect(get(values, 'commentsIds.1')).toBe(values.commentsIds[1]);
    expect(get(values, 'comments')).toBe(values.comments);
    expect(get(values, 'comments.0.text')).toBe(values.comments[0].text);
    expect(get(values, 'comments.1.text')).toBe(values.comments[1].text);
  });

  it('should return undefined when incorrect path', () => {
    expect(get(values, 'incorrect')).toBeUndefined();
    expect(get(values, 'incorrect.incorrect')).toBeUndefined();
    expect(get(values, 'incorrect.0.incorrect')).toBeUndefined();
    expect(get(values, 'id.incorrect')).toBeUndefined();
    expect(get(values, 'info.incorrect')).toBeUndefined();
    expect(get(values, 'commentsIds.3')).toBeUndefined();
    expect(get(values, 'commentsIds.3.incorrect')).toBeUndefined();
    expect(get(values, 'comments.3')).toBeUndefined();
    expect(get(values, 'comments.2.incorrect')).toBeUndefined();
    expect(get(values, 'comments.3.incorrect')).toBeUndefined();
  });
});
