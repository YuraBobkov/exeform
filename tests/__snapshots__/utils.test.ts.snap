// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set should set non-existent path 1`] = `
Object {
  "foo": Array [
    Object {
      "bar": "baz",
    },
  ],
}
`;

exports[`set should set non-existent path 2`] = `
Object {
  "foo": Array [
    Object {
      "bar": Object {
        "baz": "bat",
      },
    },
  ],
}
`;

exports[`set should set value 1`] = `
Object {
  "comments": Array [
    Object {
      "id": 56,
      "text": "foo",
    },
    Object {
      "id": 78,
      "text": "bar",
    },
  ],
  "commentsIds": Array [
    56,
    78,
  ],
  "id": 84,
  "info": Object {
    "name": "John",
  },
}
`;

exports[`set should set value to array 1`] = `
Object {
  "comments": Array [
    Object {
      "id": 56,
      "text": "baz",
    },
    Object {
      "id": 78,
      "text": "bar",
    },
  ],
  "commentsIds": Array [
    56,
    78,
  ],
  "id": 42,
  "info": Object {
    "name": "John",
  },
}
`;

exports[`set should set value to array 2`] = `
Object {
  "comments": Array [
    Object {
      "id": 56,
      "text": "foo",
    },
    Object {
      "id": 78,
      "text": "bat",
    },
  ],
  "commentsIds": Array [
    56,
    78,
  ],
  "id": 42,
  "info": Object {
    "name": "John",
  },
}
`;

exports[`set should set value to nested path 1`] = `
Object {
  "comments": Array [
    Object {
      "id": 56,
      "text": "foo",
    },
    Object {
      "id": 78,
      "text": "bar",
    },
  ],
  "commentsIds": Array [
    56,
    78,
  ],
  "id": 42,
  "info": Object {
    "name": "Mike",
  },
}
`;
